### PR TITLE
ARROW-6629: [Doc] [C++] Add filesystem docs

### PR DESCRIPTION
**Pull request by pitrou**
Also adds and improves other C++ doc items.

### CODE REVIEW

File: docs/source/cpp/io.rst
Comment (pitrou) [inline docs/source/cpp/io.rst]: @nealrichardson Would be nice to know if this helps.

File: docs/source/cpp/io.rst
Comment (nealrichardson) [inline docs/source/cpp/io.rst]: Is this `...` a note to self to fill this in, or is it supposed to be there?

**Review by nealrichardson — APPROVED** [review]
I haven't tested the rendered docs pages, but I'm heartily +1 on backfilling docs. 

Once CI is happy I'll merge, unless there's objection. I'm sure this could be improved further but I don't see a reason to hold it up, particularly if we can sneak it into 0.15.

File: docs/source/cpp/io.rst
Comment (pitrou) [reply docs/source/cpp/io.rst]: It's just a way of writing "etc.". Is it incorrect?

**Comment by pitrou** [comment]
Travis-CI build (green): https://travis-ci.org/pitrou/arrow/builds/589009656

AppVeyor doesn't build docs.

File: docs/source/cpp/io.rst
Comment (nealrichardson) [reply docs/source/cpp/io.rst]:
```suggestion
Parquet data, transmitting IPC streams, and more.
```

**Comment by nealrichardson** [comment]
Ok, if you would please accept my "suggestion" (I'm not authorized) or similar, I'll merge.

**Comment by pitrou** [comment]
Done.

**Comment by nealrichardson** [comment]
Merci!

**Pull request closed without merging**
